doc/neps/nep-0056-array-api-main-namespace.rst

```diff
@@ -0,0 +1,524 @@
+
+The most prominent users who will benefit from array API support are probably
+downstream libraries that want to start supporting CuPy, PyTorch, JAX, Dask, or
+other such libraries. SciPy and scikit-learn are already fairly far along the
+way of doing just that, and successfully support CuPy arrays and PyTorch
+tensors in a small part of their own APIs (that support is still marked as
+experimental).
+
+The main principle they use is that they replace the regular
+``import numpy as np`` with a utility function to retrieve the array library
+namespace from the input array. They call it ``xp``, which is effectively an
+alias to ``np`` if the input is a NumPy array, ``cupy`` for a CuPy array,
+``torch`` for a PyTorch tensor. This ``xp`` then allows writing code that works
+for all these libraries - because the array API standard is the common
+denominator. As a concrete example, this code is taken from ``scipy.cluster``:
+
+.. code:: python
+
+    def vq_py(obs, code_book, check_finite=True):
+        """Python version of vq algorithm"""
+        xp = array_namespace(obs, code_book)
+        obs = as_xparray(obs, xp=xp, check_finite=check_finite)
+        code_book = as_xparray(code_book, xp=xp, check_finite=check_finite)
+
+        if obs.ndim != code_book.ndim:
+            raise ValueError("Observation and code_book should have the same rank")
+
+        if obs.ndim == 1:
+            obs = obs[:, xp.newaxis]
+            code_book = code_book[:, xp.newaxis]
+
+        # Once `cdist` has array API support, this `xp.asarray` call can be removed
+        dist = xp.asarray(cdist(obs, code_book))
+        code = xp.argmin(dist, axis=1)
+        min_dist = xp.min(dist, axis=1)
+        return code, min_dist
+
+It mostly looks like normal NumPy code, but will run with for example PyTorch
+tensors as input and then return PyTorch tensors. There is a lot more to this
+story of course then this basic example. These blog posts on scikit-learn [1]_
+and SciPy's [2]_ experiences and impact (large performance gains in some cases
+- ``LinearDiscriminantAnalysis.fit`` showed ~28x gain with PyTorch on GPU vs.
+NumPy) paint a more complete picture.
+
+For end users who are using NumPy directly, little changes aside from there
+being fewer differences between NumPy and other libraries they may want to use
+as well. This shortens their learning curve and makes it easier to switch
+between NumPy and PyTorch/JAX/CuPy. In addition, they should benefit from
+array-consuming libraries starting to support multiple array libraries, making
+their experience of using a stack of Python packages for scientific computing
+or data science more seamless.
+
+Finally, for authors of other array libraries as well as tools like Numba,
+array API standard support should save them time. The design rules ([3]_), and
+in some cases new APIs like the ``unique_*`` ones, are easier to implement on
+GPU and for JIT compilers as a result of more predictable behavior.
+
```

Comment:
Perhaps this is a bit too speculative or big-picture, but I think a comment about the potential impact on new (in the sense of not done before with any array library) functionality - only needing to have 1 array-agnostic implementation, rather than 5+ similar-but-incompatible implementations for each array library ecosystem - wouldn't be out of place while we are talking about consumer libraries.

Reply:
If you don't mind, I'd like to leave this as is, since it's described in the array API standard language, in NEP 47, and in blog posts, talks, etc. - I'm running short on time for 2.0, and there's only so many times I can rewrite the same abstract point. NEP 47 has for example:

_For a longer discussion on the purpose of the array API standard we refer to the [Purpose and Scope section of the array API standard](https://data-apis.github.io/array-api/latest/purpose_and_scope.html) and the two blog posts announcing the formation of the Consortium [[1]](file:///home/rgommers/code/numpy/doc/neps/_build/html/nep-0047-array-api-standard.html#id6) and the release of the first draft version of the standard for community review [[2]](file:///home/rgommers/code/numpy/doc/neps/_build/html/nep-0047-array-api-standard.html#id7)._

Reply:
Sure thing!